cmd/adminapi/main.go

```diff
@@ -74,14 +74,23 @@ func realMain(ctx context.Context) error {
 	// Setup monitoring
 	logger.Info("configuring observability exporter")
 	oeConfig := cfg.ObservabilityExporterConfig()
-	oe, err := observability.NewFromEnv(ctx, oeConfig)
+	// Create a separate ctx to allow metric exporter to finish uploading the
+	// last batch of metrics when SIGTERM is received.
+	metricCtx, cancel := context.WithCancel(context.Background())
+	oe, err := observability.NewFromEnv(metricCtx, oeConfig)
 	if err != nil {
+		cancel()
 		return fmt.Errorf("unable to create ObservabilityExporter provider: %w", err)
 	}
 	if err := oe.StartExporter(); err != nil {
+		cancel()
 		return fmt.Errorf("error initializing observability exporter: %w", err)
 	}
-	defer oe.Close()
+	defer func() {
```

Comment:
Sorry, but I don't like this API or the repeatability. It's really too easy for someone to miss a `cancel` or misuse this. Can we push some of this logic into the observability package? Or can we just create a new context (not cancelable) and leave flow control timeouts up to Cloud Run?

```go
metricsCtx := context.Background()
oe, err := observability.NewFromEnv(metricCtx, oeConfig)
if err != nil {
  return ...
}
defer oe.Close()
```

That will still give the oe the cleanup signal and it will prevent the full function from stopping until the defer finishes (assuming `Close` is blocking).

Even better, we could drop the context parameter from the o11y entirely and just construct the context internally. In both cases, we're losing the logger that's attached to the provided context, so maybe we want `NewFromEnv` to look like this:

```go
NewFromEnv(ctx context.Context, config *Config) {
  logger := logger.FromContext(ctx)

  ctx = context.Background()
  ctx = logging.WithLogger(ctx, logger)

  // Existing stuffs
}
```

Then all the `main.go` functions are entirely unchanged.

Reply:
Updated with the simplified logic. The exporter will use its own ctx anyway so `observability.NewFromEnv` don't need to take a ctx.